### ISSUE TITLE
eval in a (nonclosure) sub sees lexicals outside the sub

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#2108]
Migrated from [rt.perl.org#3398](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=3398) (status was 'resolved')

Searchable as RT3398$

**Comment by p5pRT** [comment]
### From @ysth

#### Created by @ysth



Here's a simple example&#8203;:

\[D&#8203;:\\home\\sthoenna\]perl \-wl
$x = $x = 42;
\{ my $x; sub incx \{ eval '\+\+$x' \} \}
print incx;
print incx;
\_\_END\_\_
1
2

Here\, the eval modifies the lexical $x\, where IMO it should modify the
package $x\.  I think eval should never reach outside a sub to pick up
lexicals \(unless of course the sub is explicitly a closure on those
lexicals\)\.  Here is another example where this causes highly
unintuitive results&#8203;:

\[D&#8203;:\\home\\sthoenna\]perl \-wl
$x = $x = 42;
sub a \{ my $x; sub incx \{ eval '\+\+$x' \}; \+\+$x \}
print incx;
print a;
print incx;
\_\_END\_\_
1
2
1

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.6.0:

Configured by sthoenna at Thu Mar 23 16:36:26 PST 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=os2, osvers=2.30, archname=os2
    uname='os2 efn.org 2 2.30 i386  '
    config_args='-de -Dprefix=d:/perl -Aoptimize=-DDEBUGGING'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define 
    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc='gcc', optimize='-O2 -fomit-frame-pointer -malign-loops=2 -malign-jumps=2 -malign-functions=2 -s -DDEBUGGING', gccversion=2.8.1
    cppflags='-Zomf -Zmt -DDOSISH -DOS2=2 -DEMBED -I. -D_EMX_CRT_REV_=62'
    ccflags ='-Zomf -Zmt -DDOSISH -DOS2=2 -DEMBED -I. -D_EMX_CRT_REV_=62'
    stdchar='char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=4
    alignbytes=4, usemymalloc=y, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags ='-Zexe -Zomf -Zmt -Zcrtdll -Zstack 32000'
    libpth=d:/emx/lib d:/emx/lib/mt
    libs=-lsocket -lm -lbsd
    libc=d:/emx/lib/mt/c_import.lib, so=dll, useshrplib=true, libperl=libperl.lib
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
    cccdlflags='-Zdll', lddlflags='-Zdll -Zomf -Zmt -Zcrtdll -s'

Locally applied patches:
    


@INC for perl v5.6.0:
    d:/perl/lib/5.6.0/os2
    d:/perl/lib/5.6.0
    d:/perl/lib/site_perl/5.6.0/os2
    d:/perl/lib/site_perl/5.6.0
    d:/perl/lib/site_perl
    .


Environment for perl v5.6.0:
    HOME=d:\home\sthoenna
    LANG=en_us
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=d:\bin;C:\OS2;d:\perl\bin;C:\OS2\SYSTEM;C:\OS2\INSTALL;C:\;C:\OS2\MDOS;C:\OS2\APPS;C:\MMOS2;d:\os2apps\util;d:\DOSAPPS\UTIL;c:\sio;D:\WINDOWS;d:\pdksh;d:\emx\bin;d:\emacs\19.33\bin;d:\ispell
    PERL_BADLANG (unset)
    PERL_SH_DIR=d:\BIN
    SHELL (unset)


```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]


sthoenna@&#8203;efn\.org wrote&#8203;:
> 
> \[D&#8203;:\\home\\sthoenna\]perl \-wl
> $x = $x = 42;
> \{ my $x; sub incx \{ eval '\+\+$x' \} \}
> print incx;
> print incx;
> \_\_END\_\_
> 1
> 2

Here is Ilya's excellent explanation of this from clpmisc&#8203;:

&nbsp;    http&#8203;://www\.deja\.com/\[ST\_rn=ps\]/getdoc\.xp?AN=545654855&fmt=text

I think he mentioned later that a warning might be possible\.

> Here\, the eval modifies the lexical $x\, where IMO it should modify the
> package $x\.  I think eval should never reach outside a sub to pick up
> lexicals \(unless of course the sub is explicitly a closure on those
> lexicals\)\.  Here is another example where this causes highly
> unintuitive results&#8203;:
> 
> \[D&#8203;:\\home\\sthoenna\]perl \-wl
> $x = $x = 42;
> sub a \{ my $x; sub incx \{ eval '\+\+$x' \}; \+\+$x \}
> print incx;
> print a;
> print incx;
> \_\_END\_\_
> 1
> 2
> 1

The only unintuitive thing here may be the 2\.  But only if you think

&nbsp;    my $foo if 0;

is unintuitive\.

**Comment by p5pRT** [comment]
### From @mjdominus


A few months ago I did some research into what the Lisp folks thought
about this\, since they have some experience in this area\.  It is
considered a difficult area\, hard to decide what the right thing is\.
The designers of Scheme\, for example\, sidestepped the problem&#8203;:
Scheme's eval function has a required second parameter that explicitly
specifies the environment in which the eval will be performed\.

I have notes somewhere about the way 'eval' behaves in Common Lisp\,
and the rationale behind the design\.  I will try to find them and send
them when I get back home next week\.

**Comment by p5pRT** [comment]
### From @ysth

In article \<394D6EBE\.1F2C9AFE@&#8203;home\.com>\,
Rick Delaney \<rick\.delaney@&#8203;home\.com> wrote&#8203;:
> > \[D&#8203;:\\home\\sthoenna\]perl \-wl
> > $x = $x = 42;
> > sub a \{ my $x; sub incx \{ eval '\+\+$x' \}; \+\+$x \}
> > print incx;
> > print a;
> > print incx;
> > \_\_END\_\_
> > 1
> > 2
> > 1
> 
> The only unintuitive thing here may be the 2\.  But only if you think
> 
>     my $foo if 0;
> 
> is unintuitive\.

Yes\, it is indeed the 2 to which I was referring\.  There is no excuse
for C\<my $x;> being any different than C\<my $x=undef;>\.

You can choose to call it a bug due to the how lexical scope is
implemented \(with the lexical reset to undef upon leaving the scope\,
wrongly trusting that it will stay that way until the scope is entered
again\)\.  Or you can say\, as I do\, that code in a sub has no business
messing with outside lexicals unless it is a closure\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]


Yitzchak Scott\-Thoennes wrote&#8203;:
> 
> You can choose to call it a bug due to the how lexical scope is
> implemented \(with the lexical reset to undef upon leaving the scope\,
> wrongly trusting that it will stay that way until the scope is entered
> again\)\.  

Which I do\.

> Or you can say\, as I do\, that code in a sub has no business
> messing with outside lexicals unless it is a closure\.

Meaning that these lexicals should be read\-only?  Or should code like
this \(admittedly absurd example\) no longer have access to the outside
lexical?

&nbsp;    my $Pseudo\_global = 2;

&nbsp;    sub double \{
&nbsp;        my \($x\) = @&#8203;\_;
&nbsp;        eval '$x \* $Pseudo\_global';
&nbsp;    \}

As long as eval has access to lexicals in its scope I think it should be
able to modify those lexicals\, regardless of whether it is in a sub or
not\.

**Comment by p5pRT** [comment]
### From @ysth

In article \<394F69AC\.663723FC@&#8203;home\.com>\,
Rick Delaney \<rick\.delaney@&#8203;home\.com> wrote&#8203;:
> 
> Yitzchak Scott\-Thoennes wrote&#8203;:
> > 
> > You can choose to call it a bug due to the how lexical scope is
> > implemented \(with the lexical reset to undef upon leaving the scope\,
> > wrongly trusting that it will stay that way until the scope is entered
> > again\)\.  
> 
> Which I do\.
> 
> > Or you can say\, as I do\, that code in a sub has no business
> > messing with outside lexicals unless it is a closure\.
> 
> Meaning that these lexicals should be read\-only?

No\, they should not be used at all\.

> Or should code like this \(admittedly absurd example\) no longer have
> access to the outside lexical?
> 
>     my $Pseudo\_global = 2;
> 
>     sub double \{
>         my \($x\) = @&#8203;\_;
>         eval '$x \* $Pseudo\_global';
>     \}
> 
> As long as eval has access to lexicals in its scope I think it should be
> able to modify those lexicals\, regardless of whether it is in a sub or
> not\.

If you want double to be a closure on $Psuedo\_global\, make it one by
using the variable where the compiler can see it\.  If you don't wan't
it to be a closure\, it should be using $&#8203;::Psuedo\_global\.

**Comment by p5pRT** [comment]
### From @btilly

Yitzchak Scott\-Thoennes wrote&#8203;:
>Rick Delaney \<rick\.delaney@&#8203;home\.com> wrote&#8203;:
\[\.\.\.\]
> > > Or you can say\, as I do\, that code in a sub has no business
> > > messing with outside lexicals unless it is a closure\.
> >
> > Meaning that these lexicals should be read\-only?
>
>No\, they should not be used at all\.

Why not?

You say it should be a closure before you do it\.  I say that it is
\*always\* a closure\.  A closure is what you get when you have a lexical
inside of a block\, and perlsyn very clearly spells out what a block is\.
Do you think that the definition of a block needs to be changed?

> > Or should code like this \(admittedly absurd example\) no longer have
> > access to the outside lexical?
> >
> >     my $Pseudo\_global = 2;
> >
> >     sub double \{
> >         my \($x\) = @&#8203;\_;
> >         eval '$x \* $Pseudo\_global';
> >     \}
> >
> > As long as eval has access to lexicals in its scope I think it should be
> > able to modify those lexicals\, regardless of whether it is in a sub or
> > not\.
>
>If you want double to be a closure on $Psuedo\_global\, make it one by
>using the variable where the compiler can see it\.  If you don't wan't
>it to be a closure\, it should be using $&#8203;::Psuedo\_global\.

Are you advocating breaking existing code that happens to rely on a
behaviour that has worked for ages and has been documented for ages as
being how Perl works?  I certainly hope not\!

Currently the compiler has no problems with the above code\, and the
above code should work according to the documentation\.

But I am curious\.  How do you think that block and closure should be
operationally defined?  Keep in mind that I am going to be very
unhappy if things like this silly example do not work\.\.\.

&nbsp;    sub create\_stuff \{
&nbsp;        my $foo = shift;
&nbsp;        if \(&some\_test\) \{
&nbsp;            my $sub\_name = shift;
&nbsp;            eval "sub $sub\_name \{&process\(\\$foo\, \\@&#8203;\_\)\}";
&nbsp;        \}
&nbsp;    \}

\(The point being that evals very frequently have legitimate reason to
see lexical variables a few blocks up\.\.\.\)

Regards\,
Ben
\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_
Get Your Private\, Free E\-mail from MSN Hotmail at http&#8203;://www\.hotmail\.com

**Comment by p5pRT** [comment]
### From @ysth


In article \<20000620202329\.79061\.qmail@&#8203;hotmail\.com>\,
"Ben Tilly" \<ben\_tilly@&#8203;hotmail\.com> wrote&#8203;:
> Yitzchak Scott\-Thoennes wrote&#8203;:
> >Rick Delaney \<rick\.delaney@&#8203;home\.com> wrote&#8203;:
> \[\.\.\.\]
> > > > Or you can say\, as I do\, that code in a sub has no business
> > > > messing with outside lexicals unless it is a closure\.
> > >
> > > Meaning that these lexicals should be read\-only?
> >
> >No\, they should not be used at all\.
> 
> Why not?
> 
> You say it should be a closure before you do it\.  I say that it is
> \*always\* a closure\.  A closure is what you get when you have a lexical
> inside of a block\, and perlsyn very clearly spells out what a block is\.
> Do you think that the definition of a block needs to be changed?

You may look at it and recognize it as a closure\, and I may look at it
and recognize it as a closure\, and Perl may document it as a closure\,
but \*p\*erl doesn't \(and \*can't\*\) know it is a closure\.

> > > Or should code like this \(admittedly absurd example\) no longer have
> > > access to the outside lexical?
> > >
> > >     my $Pseudo\_global = 2;
> > >
> > >     sub double \{
> > >         my \($x\) = @&#8203;\_;
> > >         eval '$x \* $Pseudo\_global';
> > >     \}
> > >
> > > As long as eval has access to lexicals in its scope I think it should be
> > > able to modify those lexicals\, regardless of whether it is in a sub or
> > > not\.
>
> >If you want double to be a closure on $Psuedo\_global\, make it one by
> >using the variable where the compiler can see it\.  If you don't wan't
> >it to be a closure\, it should be using $&#8203;::Psuedo\_global\.
> 
> Are you advocating breaking existing code that happens to rely on a
> behaviour that has worked for ages and has been documented for ages as
> being how Perl works?  I certainly hope not\!

The above code certainly doesn't 'work' by my definition of 'work'&#8203;:
\[D&#8203;:\\home\\sthoenna\]perl \-wl
\{
&nbsp;   my $Psuedo\_global = 2;

&nbsp;   sub double \{
&nbsp;      my \($x\) = @&#8203;\_;
&nbsp;      eval '$x \* $Psuedo\_global';
&nbsp;   \}

&nbsp;   print double\(3\);
\}
print double\(3\);
\_\_END\_\_
6
Use of uninitialized value in multiplication \(\*\) at \(eval 2\) line 1\.
0

If you truly want double to be a closure\, do it this way&#8203;:
\[D&#8203;:\\home\\sthoenna\]perl \-wl
\{
&nbsp;   my $Psuedo\_global = 2;
&nbsp;   sub double \{
&nbsp;      my \($x\) = @&#8203;\_;
&nbsp;      return eval '$x \* $Psuedo\_global';
&nbsp;      $Psuedo\_global if 0;
&nbsp;   \}
&nbsp;   print double\(3\);
\}
print double\(3\);
\_\_END\_\_
6
6

Nevertheless\, I concede the point\.  You are right that the above code
should probably continue to work where it does now\.

> Currently the compiler has no problems with the above code\, and the
> above code should work according to the documentation\.
> 
> But I am curious\.  How do you think that block and closure should be
> operationally defined?  Keep in mind that I am going to be very
> unhappy if things like this silly example do not work\.\.\.
> 
>     sub create\_stuff \{
>         my $foo = shift;
>         if \(&some\_test\) \{
>             my $sub\_name = shift;
>             eval "sub $sub\_name \{&process\(\\$foo\, \\@&#8203;\_\)\}";
>         \}
>     \}
> 
> \(The point being that evals very frequently have legitimate reason to
> see lexical variables a few blocks up\.\.\.\)

I wouldn't call this case silly\.  It is the paradigm for what is probably
one of the top two uses of closures\.

But it's not the case I have a problem with\!  Here\, when the sub is compiled\,
the compiler sees that $foo is used and the sub is a true closure\.

I give up\.  I don't see a way to make your 'double' case above work\, and
my case work&#8203;:

\[D&#8203;:\\home\\sthoenna\]perl \-wl
$x = $x = 42;
sub a \{ my $x; sub incx \{ eval '\+\+$x' \}; \+\+$x \}
print incx;
print a;
print incx;
\_\_END\_\_
1
2
1

unless there is a way to retroactively make incx recognize upon being
exited from that it really was a closure on $x\, and have it grab the
current $x and allocate a new one for the next time a\(\) is entered\.

But that sounds like a pipe dream\.\.\.

In article \<20000619034516\.1245\.qmail@&#8203;plover\.com>\,
Mark\-Jason Dominus \<mjd@&#8203;plover\.com> wrote&#8203;:
> 
> A few months ago I did some research into what the Lisp folks thought
> about this\, since they have some experience in this area\.  It is
> considered a difficult area\, hard to decide what the right thing is\.
> The designers of Scheme\, for example\, sidestepped the problem&#8203;:
> Scheme's eval function has a required second parameter that explicitly
> specifies the environment in which the eval will be performed\.
> 
> I have notes somewhere about the way 'eval' behaves in Common Lisp\,
> and the rationale behind the design\.  I will try to find them and send
> them when I get back home next week\.

Looking forward to seeing it\.

**Comment by p5pRT** [comment]
### From @btilly

Yitzchak Scott\-Thoennes wrote&#8203;:
>
>In article \<20000620202329\.79061\.qmail@&#8203;hotmail\.com>\,
>"Ben Tilly" \<ben\_tilly@&#8203;hotmail\.com> wrote&#8203;:
\[\.\.\.\]
>You may look at it and recognize it as a closure\, and I may look at it
>and recognize it as a closure\, and Perl may document it as a closure\,
>but \*p\*erl doesn't \(and \*can't\*\) know it is a closure\.

Documentation explains how you are supposed to think about it\.  Of
course how you think about it and what the program really does
will never exactly match\.\.\.

\[\.\.\.\]
> > Are you advocating breaking existing code that happens to rely on a
> > behaviour that has worked for ages and has been documented for ages as
> > being how Perl works?  I certainly hope not\!
>
>The above code certainly doesn't 'work' by my definition of 'work'&#8203;:
>\[D&#8203;:\\home\\sthoenna\]perl \-wl
>\{
>    my $Psuedo\_global = 2;
>
>    sub double \{
>       my \($x\) = @&#8203;\_;
>       eval '$x \* $Psuedo\_global';
>    \}
>
>    print double\(3\);
>\}
>print double\(3\);
>\_\_END\_\_
>6
>Use of uninitialized value in multiplication \(\*\) at \(eval 2\) line 1\.
>0

That is a nasty gotcha\.  I don't like it either\.  But I would not
want to have anonymous functions being produced without being able
to clean up the temporary variables used to produce them either\.

>If you truly want double to be a closure\, do it this way&#8203;:
>\[D&#8203;:\\home\\sthoenna\]perl \-wl
>\{
>    my $Psuedo\_global = 2;
>    sub double \{
>       my \($x\) = @&#8203;\_;
>       return eval '$x \* $Psuedo\_global';
>       $Psuedo\_global if 0;
>    \}
>    print double\(3\);
>\}
>print double\(3\);
>\_\_END\_\_
>6
>6
>
>Nevertheless\, I concede the point\.  You are right that the above code
>should probably continue to work where it does now\.

Now that I see the case that you complain about\, I see your point
as well\.  OTOH you could make an argument that the above is due
to programmer error and the correct fix is this&#8203;:

\{
&nbsp;    my $Psuedo\_global = 2;

&nbsp;    sub double \{
&nbsp;       my \($x\) = @&#8203;\_;
&nbsp;       eval \{$x \* $Psuedo\_global\};
&nbsp;    \}

&nbsp;    print double\(3\);
\}
print double\(3\);

Why is this better?  Well it is a lot faster and it is much less
likely to break in the future on a buggy optimization\.

> > \[\.\.\.\] I am going to be very
> > unhappy if things like this silly example do not work\.\.\.
> >
> >     sub create\_stuff \{
> >         my $foo = shift;
> >         if \(&some\_test\) \{
> >             my $sub\_name = shift;
> >             eval "sub $sub\_name \{&process\(\\$foo\, \\@&#8203;\_\)\}";
> >         \}
> >     \}
> >
> > \(The point being that evals very frequently have legitimate reason to
> > see lexical variables a few blocks up\.\.\.\)
>
>I wouldn't call this case silly\.  It is the paradigm for what is probably
>one of the top two uses of closures\.

Well the code clearly didn't belong to a real program\, and the
reason that I called it silly was that I looked for a silly
reason to want to call eval inside of an if\.  In general I prefer

>But it's not the case I have a problem with\!  Here\, when the sub is 
>compiled\,
>the compiler sees that $foo is used and the sub is a true closure\.
>
>I give up\.  I don't see a way to make your 'double' case above work\, and
>my case work&#8203;:
>
>\[D&#8203;:\\home\\sthoenna\]perl \-wl
>$x = $x = 42;
>sub a \{ my $x; sub incx \{ eval '\+\+$x' \}; \+\+$x \}
>print incx;
>print a;
>print incx;
>\_\_END\_\_
>1
>2
>1
>
>unless there is a way to retroactively make incx recognize upon being
>exited from that it really was a closure on $x\, and have it grab the
>current $x and allocate a new one for the next time a\(\) is entered\.

Heh\.  You have far worse problems than that there\!

sub remember \{
&nbsp;    my $x = shift;
&nbsp;    sub incr \{
&nbsp;        \+\+$x;
&nbsp;    \}
\}
&remember\(5\);
print &incr;
&remember\(5\);
print &incr;

This won't do what you want\, and warnings explains why not\.  Now
an argument can be made that this should have been written as
follows&#8203;:

\{
&nbsp;    my $x;
&nbsp;    sub remember \{$x = shift;\}
&nbsp;    sub incr \{\+\+$x\}
\}

OTOH the mod\_perl people have a case where the first really should
be supported\.  \(Look for mod\_perl debugging issues\.\.\.\)

>But that sounds like a pipe dream\.\.\.

Perhaps a better answer would be to have certain globally visible
constructs internal to a block \*force\* that block to never undergo
reinitializations\.  The only candidate that I have for this is
a global subroutine internal to the block\.  That would make both
your desired eval and the mod\_perl cases work like you and they
want\.  However anonymous subroutines created inside of the block
most emphatically should not cause any such effect\.

I suspect that this would not be a minor change though\.\.\.

Cheers\,
Ben
\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_
Get Your Private\, Free E\-mail from MSN Hotmail at http&#8203;://www\.hotmail\.com
